crates/binstalk/src/helpers/git.rs

```diff
@@ -0,0 +1,80 @@
+use std::{num::NonZeroU32, path::Path, str::FromStr, sync::atomic::AtomicBool};
+
+use gix::{clone, create, open, progress::Discard, remote, Url};
+use thiserror::Error as ThisError;
+use tracing::debug;
+
+#[derive(Debug, ThisError)]
+#[non_exhaustive]
+pub enum GitError {
+    #[error("Failed to prepare for fetch: {0}")]
+    PrepareFetchError(#[source] Box<clone::Error>),
+
+    #[error("Failed to fetch: {0}")]
+    FetchError(#[source] Box<clone::fetch::Error>),
+
+    #[error("Failed to checkout: {0}")]
+    CheckOutError(#[source] Box<clone::checkout::main_worktree::Error>),
+}
+
+impl From<clone::Error> for GitError {
+    fn from(e: clone::Error) -> Self {
+        Self::PrepareFetchError(Box::new(e))
+    }
+}
+
+impl From<clone::fetch::Error> for GitError {
+    fn from(e: clone::fetch::Error) -> Self {
+        Self::FetchError(Box::new(e))
+    }
+}
+
+impl From<clone::checkout::main_worktree::Error> for GitError {
+    fn from(e: clone::checkout::main_worktree::Error) -> Self {
+        Self::CheckOutError(Box::new(e))
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct GitUrl(Url);
+
+impl FromStr for GitUrl {
+    type Err = gix::url::parse::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Url::try_from(s).map(Self)
+    }
+}
+
+#[derive(Debug)]
+pub struct Repository(gix::Repository);
+
+impl Repository {
+    /// WARNING: This is a blocking operation, if you want to use it in
+    /// async context then you must wrap the call in [`tokio::task::spawn_blocking`].
+    pub fn shallow_clone(url: GitUrl, path: &Path) -> Result<Self, GitError> {
+        debug!("Shallow cloning {url:?} to {}", path.display());
+
+        let mut progress = Discard;
```

Comment:
Ideally we should log progress using `tracing`, but the `gix::Progress` is so complicated.

@passcod Do you have any idea on how to log progress?

Reply:
Ideally:
- in gui/tty mode, using one of the progress crates to render a spinner or bar
- otherwise (e.g. for CI use), log on every 10% or something like curl does (i think git does similar)

Reply:
That sounds good, do you know any crate for translating `Progress` to `tracing`/`log` calls?

I've found [`prodash`](https://docs.rs/prodash) for GUI/TTY interface, though it also seems to be a bit overblown for our use case.

Reply:
`cargo` has the same problem to solve and in it has [a function to help with obtaining progress](https://github.com/rust-lang/cargo/blob/1a498b6c1c119a79d677553862bffae96b97ad7f/src/cargo/sources/git/oxide.rs#L63). It will adapt progress obtained from the [`progress root`](https://github.com/rust-lang/cargo/blob/1a498b6c1c119a79d677553862bffae96b97ad7f/src/cargo/sources/git/oxide.rs#L37-L43) in a separate thread to the `cargo` progress bar. Something similar would have to be done here if a `prodash` renderer isn't chosen, and not doing so is understandable as it's made to show everything of what's going on to help the developer, and even for these it can be 'too much', speaking from my own experience. 

There is [a `log` renderer](https://docs.rs/prodash/25.0.0/prodash/progress/struct.Log.html) as well, and it emits messages using the `log` crate.

Hope that helps.

Reply:
Thank you for your help @Byron !
I will implement `log` renderer before trying out the `cargo` progress bar solution.

Reply:
Ok, I've implemented `tracing` progress.
It looks quite good to me.